src/base_home.h

```diff
@@ -0,0 +1,201 @@
+#pragma once
+#ifndef BASE_HOME_H
+#define BASE_HOME_h
+
+#include "submap.h"
+#include "ammo.h"
+#include "string_id.h"
+
+#include <unordered_map>
+#include <set>
+#include <list>
+
+struct submap;
+struct tripoint;
+class ammunition_type;
+
+//Furnature accepted as a storage place.
+//TODO: Move to json once I get around to figure out how cdda's json works
+const std::set<furn_id> store_furn {
+    "f_locker",
+    "f_dresser",
+    "f_rack",
+    "f_bookcase",
+    "f_safe_o"
+}
+
+//Need way to dermine if it counts as a bunk.
+//Finding the "This seems like a nice place to sleep" code is annoying so,
+//lets do the same should be in json hack for now
+const std::set<furn_id> sleep_furn {//Furnature that counts as a bunk
+    "f_bed",
+    "f_makeshift_bed",
+    "f_straw_bed"
+}
+const std::set<trap_id> sleep_trap {//Traps that count as a bunk
+    "tr_cot",
+    "tr_rollmat",
+    "tr_fur_rollmat"
+}
+
+
+/**
+ * Flags unique to a base.
+ */
+enum base_flag
+{
+    BASE_IN,   /**Inside of actual base*/
+    BASE_GND,  /**On base grounds. i.e. between perimeter fence and building*/
+    BASE_WALL, /**The outer walls of the base building.  Separates BASE_IN and BASE_GND*/
+    BASE_EDGE, /**Outer edge of base area (perimeter wall/fence).  Used when base area extends beyond the building structure.*/
+    NO_NPC,    /**NPCs stay out.  Friendly NPCs won't travel through tile with this flag.*/
+    COMMUNAL,  /**Designated for communal use.  NPCs take and leave items here.*/
+    OWNED_P,   /**Claimed by the player.  Friendly NPCs won't take from or use anything on this tile.*/
+    OWNED_N,   /**Claimed by an NPC.  Takeing or useing stuff here will earn their anger.  Not used by other NPCs.*/
+};
+
+//furniture flags in json
+//BASE_CORE -> Is the very heart and soul of a base.
+//BASE_CMD# -> Level provided by control core.
+//SYS_LINK -> Acts as a data conduit for the command system.  If adjacent or forms a path to BASE_CORE, allows placement of aux systems.
+//AUX_SYS -> Is an auxiliary system. Must be built adjacent to BASE_CORE flag or CMD_SYS flag linked to a BASE_CORE flag.
+//CONNECTABLE -> Possible to install network module
+//NETWORKED -> Network enabled; able to form connection between command and remote system.
+
+enum class ration_enum : char {
+    Not Rationed = 0,
+    Lightly Rationed = 1,
+    Moderately Rationed = 2,
+    Heavily Rationing = 3
+}
+
+/**
+ * Contain's entire base map.  Point is to merge base's submaps into one object.
```

Comment:
Operating directly on submaps by a class that isn't `map` is a huge red flag that you're doing something wrong.
This will cause a lot of trouble and you really don't want to do that.

I'm not sure how you want to process the camps, but it would be much safer to do this:
* Remember the camps (all of them in given area) in `overmap` (not submap)
* In a camp, note its `tripoint origin` - this would be the submap coordinate of its upper-left corner
* When you need to process the camp, load a `tinymap` centered on camp's origin

For `tinymap` example, check `mission_start::place_dog`. For a more complex example, you can check `place_caravan_ambush`, but DON'T replicate that - placing tons of things that way is bad.

By noting the camps in overmap, you may be able to make them act even when the player isn't around.

Reply:
I don't think I have actually gotten my head around how all the map stuff actually works.  My future plans basically have two requirements with the map: being able to add an additional set of flags to tiles in its area (npc keep out on doors, ownership to furniture etc); which I have done by adding bflag, another layer to the map grid.  And, being able to to keep track of locations and get what's there (communal storage).  As an extra I also want to disable the spawn points within the base because having hulks spawn inside your base is just dumb.

Reply:
Should I change `base_home::overlay` to be a `tinymap` with the extra bflag overlay?  `gen_overlay` would then be the function call to create the tinymap.

Reply:
No
`tinymap` is for when the base does something. In your base structure, you only want to keep locations, not copy of the entire map.
For example, if you want to simulate a month of NPCs eating rations, you do this:
* Calculate NPC needs for this month
* Create a `tinymap` in base's origin
* Using `tinymap`, go through food items in the base and delete those that the NPCs "ate"
* Call `save` function of the `tinymap`
* Do NOT save the `tinymap` structure anywhere, it should be deleted after you finish this simulation

Reply:
Cool, thanks